nav2_graceful_motion_controller/src/smooth_control_law.cpp

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2023 Alberto J. Tudela Roldán
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "geometry_msgs/msg/pose_stamped.hpp"
+#include "nav2_util/geometry_utils.hpp"
+#include "nav2_graceful_motion_controller/ego_polar_coords.hpp"
+#include "nav2_graceful_motion_controller/smooth_control_law.hpp"
+
+namespace nav2_graceful_motion_controller
+{
+
+SmoothControlLaw::SmoothControlLaw(
+  double k_phi, double k_delta, double beta, double lambda, double slowdown_radius,
+  double v_linear_min, double v_linear_max, double v_angular_max)
+: k_phi_(k_phi), k_delta_(k_delta), beta_(beta), lambda_(lambda), slowdown_radius_(slowdown_radius),
+  v_linear_min_(v_linear_min), v_linear_max_(v_linear_max), v_angular_max_(v_angular_max)
+{
+}
+
+void SmoothControlLaw::setCurvatureConstants(
+  double k_phi, double k_delta, double beta, double lambda)
+{
+  k_phi_ = k_phi;
+  k_delta_ = k_delta;
+  beta_ = beta;
+  lambda_ = lambda;
+}
+
+void SmoothControlLaw::setSlowdownRadius(double slowdown_radius)
+{
+  slowdown_radius_ = slowdown_radius;
+}
+
+void SmoothControlLaw::setSpeedLimit(
+  const double v_linear_min, const double v_linear_max,
+  const double v_angular_max)
+{
+  v_linear_min_ = v_linear_min;
+  v_linear_max_ = v_linear_max;
+  v_angular_max_ = v_angular_max;
+}
+
+geometry_msgs::msg::Twist SmoothControlLaw::calculateRegularVelocity(
+  const geometry_msgs::msg::Pose & target, const geometry_msgs::msg::Pose & current,
+  const bool & backward)
+{
+  // Convert the target to polar coordinates
+  auto ego_coords = EgocentricPolarCoordinates(target, current, backward);
+  // Calculate the curvature
+  double curvature = calculateCurvature(ego_coords.r, ego_coords.phi, ego_coords.delta);
+
+  // Adjust the linear velocity as a function of the path curvature to
+  // slowdown the controller as it approaches its target
+  double v = v_linear_max_ / (1.0 + beta_ * std::pow(fabs(curvature), lambda_));
+
+  // Set some small v_min when far away from origin to promote faster
+  // turning motion when the curvature is very high
+  v = std::clamp(v, v_linear_min_, v_linear_max_);
+
+  // Slowdown when the robot is near the target to remove singularity
+  v = std::min(v_linear_max_ * (ego_coords.r / slowdown_radius_), v);
```

Comment:
This needs to be BEFORE the std::clamp - otherwise as you approach the goal you will violate v_linear_min_ - which usually means the robot comes to a full and complete stop (v_linear_min_ should be the absolute lowest velocity that the robot can actually control, less than that and the robot will stall in place)